File: src/utils/converter.ts
import JSONBig from 'json-bigint';
import { msgPackEncoder } from './msgpack-encoder';

// Create a custom JSON parser that uses BigInt for large integers
// useNativeBigInt: true uses native BigInt for values exceeding safe integer limits
// Without alwaysParseAsBig, small integers remain as regular numbers, allowing
// msgpack to use compact encodings (fixint, uint8, etc.) instead of uint64
const JSONBigNative = JSONBig({ useNativeBigInt: true });

/**
 * Extract error message from various error types
 * Handles Error instances, json-bigint error objects, and unknown types
 */
function getErrorMessage(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  // json-bigint throws objects with name, message, at, and text properties
  if (
    error !== null &&
    typeof error === 'object' &&
    'message' in error &&
    typeof (error as { message: unknown }).message === 'string'
  ) {
    return (error as { message: string }).message;
  }
  return 'Unknown error';
}

/**
 * Convert Base64-encoded msgpack data to pretty JSON string
 * Supports uint64 values by using BigInt and json-bigint
 * Preserves float representation by adding .0 suffix to whole number floats
 * 
 * DecodedFloat64 objects have _isBigNumber = true which makes json-bigint
 * output their toJSON() result unquoted, avoiding the need for string markers.
 */
export function msgpackToJson(base64String: string): string {
  try {
    // Decode base64 to binary
    const binaryString = atob(base64String);
    const bytes = new Uint8Array(binaryString.length);
    for (let i = 0; i < binaryString.length; i++) {
      bytes[i] = binaryString.charCodeAt(i);
    }

    // Decode msgpack with float detection
    const wrappedData = msgPackEncoder.decode(bytes);

    // Convert to pretty JSON - DecodedFloat64's toJSON() handles float formatting
    return JSONBigNative.stringify(wrappedData, null, 2);
  } catch (error) {
    const message = getErrorMessage(error);
    throw new Error(`Failed to convert msgpack to JSON: ${message}`);
  }
}

/**
 * Convert JSON string to Base64-encoded msgpack data
 * Supports uint64 values via json-bigint
 * Preserves float representation for numbers written with decimal points or exponents
 */
export function jsonToMsgpack(jsonString: string): string {
  try {
    // Parse JSON with float detection to preserve float representation
    const data = msgPackEncoder.parseJsonWithFloats(jsonString);

    // Encode to msgpack with custom encoder that handles Float64
    const encoded = msgPackEncoder.encode(data);

    // Convert to base64
    const binaryString = Array.from(encoded)
      .map((byte) => String.fromCharCode(byte))
      .join('');
    return btoa(binaryString);
  } catch (error) {
    const message = getErrorMessage(error);
    throw new Error(`Failed to convert JSON to msgpack: ${message}`);
  }
}

/**
 * Validate if a string is valid Base64
 */
export function isValidBase64(str: string): boolean {
  try {
    atob(str);
    return true;
  } catch {
    return false;
  }
}

/**
 * Validate if a string is valid JSON
 */
export function isValidJson(str: string): boolean {
  try {
    JSONBigNative.parse(str);
    return true;
  } catch {
    return false;
  }
}

/**
 * Convert Base64 string to hex string with space-separated bytes
 */
export function base64ToHex(base64String: string): string {
  try {
    const binaryString = atob(base64String);
    const hexBytes: string[] = [];
    for (let i = 0; i < binaryString.length; i++) {
      const byte = binaryString.charCodeAt(i);
      hexBytes.push(byte.toString(16).padStart(2, '0').toUpperCase());
    }
    return hexBytes.join(' ');
  } catch {
    throw new Error('Invalid Base64 string');
  }
}

/**
 * Convert hex string (space-separated or continuous) to Base64 string
 */
export function hexToBase64(hexString: string): string {
  // Remove all whitespace and normalize
  const cleanHex = hexString.replace(/\s+/g, '');

  if (cleanHex.length === 0) {
    return '';
  }

  if (cleanHex.length % 2 !== 0) {
    throw new Error('Hex string must have an even number of characters');
  }

  if (!/^[0-9a-fA-F]*$/.test(cleanHex)) {
    throw new Error('Invalid hex characters');
  }

  let binaryString = '';
  for (let i = 0; i < cleanHex.length; i += 2) {
    const byte = parseInt(cleanHex.substring(i, i + 2), 16);
    binaryString += String.fromCharCode(byte);
  }

  return btoa(binaryString);
}

/**
 * Validate if a string is valid hex format
 */
export function isValidHex(str: string): boolean {
  const cleanHex = str.replace(/\s+/g, '');
  if (cleanHex.length === 0) {
    return true; // Empty is valid
  }
  if (cleanHex.length % 2 !== 0) {
    return false;
  }
  return /^[0-9a-fA-F]*$/.test(cleanHex);
}

File: src_rust/converter.rs
use base64::Engine;
use serde_json::Value;

/// Convert Base64-encoded msgpack data to pretty JSON string
/// Supports uint64 values by using serde_json's arbitrary precision feature
pub fn msgpack_to_json(base64_string: &str) -> Result<String, String> {
    // Decode base64 to binary
    let bytes = base64::engine::general_purpose::STANDARD
        .decode(base64_string)
        .map_err(|e| format!("Failed to decode base64: {}", e))?;

    // Decode msgpack to serde_json::Value
    let value: Value = rmp_serde::from_slice(&bytes)
        .map_err(|e| format!("Failed to decode msgpack: {}", e))?;

    // Convert to pretty JSON
    serde_json::to_string_pretty(&value)
        .map_err(|e| format!("Failed to serialize JSON: {}", e))
}

/// Convert JSON string to Base64-encoded msgpack data
pub fn json_to_msgpack(json_string: &str) -> Result<String, String> {
    // Parse JSON to serde_json::Value
    let value: Value = serde_json::from_str(json_string)
        .map_err(|e| format!("Failed to parse JSON: {}", e))?;

    // Encode to msgpack
    let bytes = rmp_serde::to_vec(&value)
        .map_err(|e| format!("Failed to encode msgpack: {}", e))?;

    // Convert to base64
    Ok(base64::engine::general_purpose::STANDARD.encode(&bytes))
}

/// Convert Base64 string to hex string with space-separated bytes
pub fn base64_to_hex(base64_string: &str) -> Result<String, String> {
    let bytes = base64::engine::general_purpose::STANDARD
        .decode(base64_string)
        .map_err(|_| "Invalid Base64 string".to_string())?;

    Ok(bytes
        .iter()
        .map(|b| format!("{:02X}", b))
        .collect::<Vec<_>>()
        .join(" "))
}

/// Convert hex string (space-separated or continuous) to Base64 string
pub fn hex_to_base64(hex_string: &str) -> Result<String, String> {
    // Remove all whitespace
    let clean_hex: String = hex_string.chars().filter(|c| !c.is_whitespace()).collect();

    if clean_hex.is_empty() {
        return Ok(String::new());
    }

    if clean_hex.len() % 2 != 0 {
        return Err("Hex string must have an even number of characters".to_string());
    }

    // Validate hex characters
    if !clean_hex.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err("Invalid hex characters".to_string());
    }

    // Convert hex to bytes
    let bytes: Vec<u8> = (0..clean_hex.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&clean_hex[i..i + 2], 16).unwrap())
        .collect();

    Ok(base64::engine::general_purpose::STANDARD.encode(&bytes))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_msgpack_to_json_simple() {
        // {"hello": "world"} in msgpack, base64 encoded
        let msgpack_base64 = "gaVoZWxsb6V3b3JsZA==";
        let json = msgpack_to_json(msgpack_base64).unwrap();
        let parsed: Value = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed["hello"], "world");
    }

    #[test]
    fn test_json_to_msgpack_roundtrip() {
        let json = r#"{"hello": "world"}"#;
        let msgpack = json_to_msgpack(json).unwrap();
        let back_to_json = msgpack_to_json(&msgpack).unwrap();
        let parsed: Value = serde_json::from_str(&back_to_json).unwrap();
        assert_eq!(parsed["hello"], "world");
    }

    #[test]
    fn test_base64_to_hex() {
        let hex = base64_to_hex("aGVsbG8=").unwrap();
        assert_eq!(hex, "68 65 6C 6C 6F");
    }

    #[test]
    fn test_hex_to_base64() {
        let base64 = hex_to_base64("68 65 6C 6C 6F").unwrap();
        assert_eq!(base64, "aGVsbG8=");
    }

    #[test]
    fn test_roundtrip_base64_hex() {
        let original = "gaVoZWxsb6V3b3JsZA==";
        let hex = base64_to_hex(original).unwrap();
        let result = hex_to_base64(&hex).unwrap();
        assert_eq!(result, original);
    }
}
